src/java/org/apache/cassandra/index/sai/utils/RangeIterator.java

```diff
@@ -110,6 +116,28 @@ public final void skipTo(PrimaryKey nextToken)
      */
     protected abstract void performSkipTo(PrimaryKey nextKey);
 
+    protected final IntersectionResult intersect(PrimaryKey nextKey)
```

Comment:
This seems very similar semantically to skipTo.  Is the difference that we need the MISS vs MATCH result to save a compareTo?  I don't understand what that's a big win, since most of the pain from compareTo is performing the deferred load

Reply:
I had exactly same concern about similarity to `skipTo`. Keeping two semantically similar methods makes the code more complex than it needs to be (immediately raises a question about the differences between those two, and which one should I use). If possible, could we just make `skipTo` return that additional information (did it skip exactly to the right position, or did it overskip)?

Reply:
One key difference is that `skipTo` is lazy and `intersect` is eager. We could easily fix that using a parameter on `skipTo`. I'll revisit this part of the PR today.

Reply:
After closer analysis, we use `skipTo` for to skip for requested token ranges. This should arguably be pushed down into the SAI index, in my opinion. It seems odd to create the iterators without the token boundaries for the query. However, that change is out of the scope of this PR, so I left `skipTo` while updating its method signature to take a `Token` instead of a primary key.